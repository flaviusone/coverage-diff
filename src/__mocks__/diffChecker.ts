export const diffChecker = () => ({
  regression: 'mockedRegression',
  files: 'mockedFiles',
  totals: { deltas: 'mockedTotals' }
});
