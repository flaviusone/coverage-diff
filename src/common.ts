/* eslint-disable no-undef */
export interface ICoverageSummary {
  lines: ICoverageInfo;
  statements: ICoverageInfo;
  functions: ICoverageInfo;
  branches: ICoverageInfo;
}

export interface ICoverageInfo {
  total: number;
  covered: number;
  skipped: number;
  pct: number; // Coverage percentage (covered/total ratio).
}

export interface IJsonSummary {
  [key: string]: ICoverageSummary;
}

export interface ICoverageDiffOutput {
  diff: IJsonSummary;
  results: string;
  regression: boolean;
}

export type Criteria = 'lines' | 'branches' | 'functions' | 'statements';

export interface IConfigOptions {
  checkCriteria?: Array<Criteria>;
  /* Fail coverage check if per-file coverage is lower */
  coverageThreshold?: number;
  /* Fail coverage check if per-file coverage decrease is lower */
  coverageDecreaseThreshold?: number;
}

export interface IDiffCheckResults {
  files: IFilesResults;
  totals: IFileResultFormat;
  diff: IJsonSummary;
  regression: boolean;
}

export interface IFilesResults {
  [key: string]: IFileResultFormat;
}

export interface IFileResultFormat {
  deltas: IFileResultFields;
  pcts: IFileResultFields;
  decreased: boolean;
}

export interface IFileResultFields {
  lines: number;
  functions: number;
  statements: number;
  branches: number;
}
