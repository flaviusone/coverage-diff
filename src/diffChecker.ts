import { objectToMap, mapToObject } from './helpers';
import { getSummaryPercentages } from './helpers';
import {
  IJsonSummary,
  IFilesResults,
  IFileResultFormat,
  Criteria
} from './common';
/**
 * Takes a json-summary formatted object (a diff) and checks if per-file
 *   coverage changed (increase/decrease).
 * Returns a pretty-formatted result and regression status.
 */
export const diffChecker = (
  diff: IJsonSummary,
  checkCriteria: Criteria[]
): { files: IFilesResults; regression: boolean } => {
  let regression = false;
  const diffMap = objectToMap(diff);
  const percentageMap: Map<string, IFileResultFormat> = new Map();
  const isBelowThreshold = (x: number) => x < 0;
  const nonZeroTest = (x: number) => x !== 0;

  // TODO total will have custom formatting in a future release. Exclude for now.
  diffMap.delete('total');

  diffMap.forEach((v, k) => {
    const percentages = getSummaryPercentages(v);

    const diffCriteria = checkCriteria.map(criteria => percentages[criteria]);

    const decreased = diffCriteria.some(isBelowThreshold);

    // Coverage decreased on a file. Set regression flag true.
    if (decreased) {
      regression = true;
    }

    // Skip unchanged files.
    if (diffCriteria.some(nonZeroTest)) {
      percentageMap.set(k, {
        deltas: {
          ...percentages
        },
        decreased
      });
    }
  });

  return {
    files: mapToObject(percentageMap),
    regression
  };
};
