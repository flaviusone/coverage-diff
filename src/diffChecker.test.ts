import { diffChecker } from './diffChecker';
import {
  fileFullCovered,
  coverageDecreased,
  onlyLinesIncreased,
  coverageNotChanged,
  noTotal
} from './summaries.fixture';
import { Criteria } from './common';

const checkCriteria: Criteria[] = [
  'lines',
  'branches',
  'functions',
  'statements'
];

describe('diffChecker', () => {
  describe('coverage increased', () => {
    it('should match snapshot', () => {
      expect(diffChecker(fileFullCovered, checkCriteria)).toMatchSnapshot();
    });
  });
  describe('no total', () => {
    it('should match snapshot', () => {
      expect(diffChecker(noTotal, checkCriteria)).toMatchSnapshot();
    });
  });
  describe('coverage decreased', () => {
    it('should match snapshot', () => {
      expect(diffChecker(coverageDecreased, checkCriteria)).toMatchSnapshot();
    });
  });
  describe('skip non changed files', () => {
    it('should match snapshot', () => {
      expect(diffChecker(coverageNotChanged, checkCriteria)).toMatchSnapshot();
    });
  });
  describe('only check lines', () => {
    it('should match snapshot', () => {
      expect(diffChecker(onlyLinesIncreased, ['lines'])).toMatchSnapshot();
    });
  });
});
