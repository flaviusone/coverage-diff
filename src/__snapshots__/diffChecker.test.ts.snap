// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diffChecker coverage decreased should match snapshot 1`] = `
Object {
  "files": Object {
    "fileA": Object {
      "decreased": true,
      "deltas": Object {
        "branches": -50,
        "functions": -50,
        "lines": -50,
        "statements": -50,
      },
    },
  },
  "regression": true,
  "totals": Object {
    "decreased": false,
    "deltas": Object {
      "branches": 50,
      "functions": 50,
      "lines": 50,
      "statements": 50,
    },
  },
}
`;

exports[`diffChecker coverage increased should match snapshot 1`] = `
Object {
  "files": Object {
    "fileA": Object {
      "decreased": false,
      "deltas": Object {
        "branches": 100,
        "functions": 100,
        "lines": 100,
        "statements": 100,
      },
    },
  },
  "regression": false,
  "totals": Object {
    "decreased": false,
    "deltas": Object {
      "branches": 100,
      "functions": 100,
      "lines": 100,
      "statements": 100,
    },
  },
}
`;

exports[`diffChecker no total should match snapshot 1`] = `
Object {
  "files": Object {
    "fileA": Object {
      "decreased": false,
      "deltas": Object {
        "branches": 100,
        "functions": 100,
        "lines": 100,
        "statements": 100,
      },
    },
  },
  "regression": false,
  "totals": Object {
    "decreased": false,
    "deltas": Object {
      "branches": 0,
      "functions": 0,
      "lines": 0,
      "statements": 0,
    },
  },
}
`;

exports[`diffChecker only check lines should match snapshot 1`] = `
Object {
  "files": Object {
    "fileA": Object {
      "decreased": false,
      "deltas": Object {
        "branches": -50,
        "functions": -50,
        "lines": 50,
        "statements": -50,
      },
    },
  },
  "regression": false,
  "totals": Object {
    "decreased": false,
    "deltas": Object {
      "branches": 50,
      "functions": 50,
      "lines": 50,
      "statements": 50,
    },
  },
}
`;

exports[`diffChecker skip non changed files should match snapshot 1`] = `
Object {
  "files": Object {},
  "regression": false,
  "totals": Object {
    "decreased": false,
    "deltas": Object {
      "branches": 100,
      "functions": 100,
      "lines": 100,
      "statements": 100,
    },
  },
}
`;
